peer/pendingheap/list_test.go

```diff
@@ -0,0 +1,613 @@
+			retainedAvailablePeerIDs:   []string{"1"},
+			retainedUnavailablePeerIDs: []string{"2"},
+			expectedAvailablePeers:     []string{"1"},
+			expectedUnavailablePeers:   []string{"2"},
+			peerListActions: []PeerListAction{
+				StartAction{},
+				UpdateAction{AddedPeerIDs: []string{"1"}},
+				UpdateAction{AddedPeerIDs: []string{"2"}},
+				ChooseAction{
+					ExpectedPeer:        "1",
+					InputContextTimeout: 20 * time.Millisecond,
+				},
+				ChooseAction{
+					ExpectedPeer:        "1",
+					InputContextTimeout: 20 * time.Millisecond,
+				},
+			},
+			expectedRunning: true,
+		},
+		{
+			msg: "remove unavailable peer",
+			retainedUnavailablePeerIDs: []string{"1"},
+			releasedPeerIDs:            []string{"1"},
+			peerListActions: []PeerListAction{
+				StartAction{},
+				UpdateAction{AddedPeerIDs: []string{"1"}},
+				UpdateAction{RemovedPeerIDs: []string{"1"}},
+				ChooseAction{
+					InputContextTimeout: 10 * time.Millisecond,
+					ExpectedErr:         newUnavailableError(context.DeadlineExceeded),
+				},
+			},
+			expectedRunning: true,
+		},
+		{
+			msg: "notify peer is now available",
+			retainedUnavailablePeerIDs: []string{"1"},
+			expectedAvailablePeers:     []string{"1"},
+			peerListActions: []PeerListAction{
+				StartAction{},
+				UpdateAction{AddedPeerIDs: []string{"1"}},
+				ChooseAction{
+					InputContextTimeout: 10 * time.Millisecond,
+					ExpectedErr:         newUnavailableError(context.DeadlineExceeded),
+				},
+				NotifyStatusChangeAction{PeerID: "1", NewConnectionStatus: peer.Available},
+				ChooseAction{ExpectedPeer: "1"},
+			},
+			expectedRunning: true,
+		},
+		{
+			msg: "notify peer is still available",
+			retainedAvailablePeerIDs: []string{"1"},
+			expectedAvailablePeers:   []string{"1"},
+			peerListActions: []PeerListAction{
+				StartAction{},
+				UpdateAction{AddedPeerIDs: []string{"1"}},
+				ChooseAction{ExpectedPeer: "1"},
+				NotifyStatusChangeAction{PeerID: "1", NewConnectionStatus: peer.Available},
+				ChooseAction{ExpectedPeer: "1"},
+			},
+			expectedRunning: true,
+		},
+		{
+			msg: "notify peer is now unavailable",
+			retainedAvailablePeerIDs: []string{"1"},
+			expectedUnavailablePeers: []string{"1"},
+			peerListActions: []PeerListAction{
+				StartAction{},
+				UpdateAction{AddedPeerIDs: []string{"1"}},
+				ChooseAction{ExpectedPeer: "1"},
+				NotifyStatusChangeAction{PeerID: "1", NewConnectionStatus: peer.Unavailable},
+				ChooseAction{
+					InputContextTimeout: 10 * time.Millisecond,
+					ExpectedErr:         newUnavailableError(context.DeadlineExceeded),
+				},
+			},
+			expectedRunning: true,
+		},
+		{
+			msg: "notify peer is still unavailable",
+			retainedUnavailablePeerIDs: []string{"1"},
+			expectedUnavailablePeers:   []string{"1"},
+			peerListActions: []PeerListAction{
+				StartAction{},
+				UpdateAction{AddedPeerIDs: []string{"1"}},
+				NotifyStatusChangeAction{PeerID: "1", NewConnectionStatus: peer.Unavailable},
```

Comment:
Non blocking, but, it would be cool to add a "NotifyPendingCountChangeAction" that would change a Peer's Pending count, and send a notification.  We could use this to validate some of the larger heap changes.